lib/reconstruct.ts

```diff
@@ -0,0 +1,73 @@
+import { types, Root, Token, Tokens, Group } from './types';
+import * as sets from './sets';
+
+const simplifications: [sets.SetFunc, string][] = [
+  [sets.words, '\\w'],
+  [sets.notWords, '\\W'],
+  [sets.ints, '\\d'],
+  [sets.notInts, '\\D'],
+  [sets.whitespace, '\\s'],
+  [sets.notWhitespace, '\\S'],
+  [sets.anyChar, '.']
+];
+
+const reduceStack = (stack: Token[]): string => stack.map(reconstruct).join('');
+
+const createAlternate = (token: Root | Group): string => {
+  if ('options' in token) {
+    return token.options.map(reduceStack).join('|');
+  } else if ('stack' in token) {
+    return reduceStack(token.stack);
+  } else {
+    throw new Error(`options or stack must be Root or Group token`);
+  }
+};
+
+export const reconstruct = (token: Tokens): string => {
+  switch (token.type) {
+    case types.ROOT:
+      return createAlternate(token);
+    case types.CHAR:
+      const c = String.fromCharCode(token.value);
+      return (/[[\]^.\/|?*+()]/.test(c) ? '\\' : '') + c;
+    case types.POSITION:
+      if (token.value === '^' || token.value === '$') {
+        return token.value;
+      } else {
+        return '\\' + token.value;
+      }
+    case types.REFERENCE:
+      return '\\' + token.value;
+    case types.SET:
+      for (const [set, simplification] of simplifications) {
+        if (JSON.stringify(set()) === JSON.stringify(token)) {
```

Comment:
While this isn’t wrong, it produces different results based on the input’s property ordering, which is kind of weird. A single-purpose element-by-element comparison would avoid that, and probably be faster, too. (And the calls to the set functions can go inside the map.)

Reply:
Resolved in latest push.

FYI I also have a file [here](https://github.com/jeswr/ret.js/blob/feature/set-simplficiation/lib/simplification-experiment.ts) that has most of the code to do simplifications such as `[012345678a9] -> [a\\d]` but I think the time complexity of it is too high to include in this library.